### PR TITLE
Add go client

### PR DESCRIPTION
**Pull request by ErshovYS**

### CODE REVIEW

**Comment by adwells-ds** [comment]
Remove `.idea` please

File: go/dataspike/dataspike.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/http"
+)
+
+const tokenHeader = "ds-api-token"
+
+type IDataspikeClient interface {
+	GetVerificationByID(string) (*Verification, error)
```
Comment (adwells-ds) [inline go/dataspike/dataspike.go:15]:
```suggestion
	GetVerificationByID(UUID) (*Verification, error)
```
`github.com/gofrs/uuid`

File: go/dataspike/dataspike.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+	GetVerificationByShortID(string) (*Verification, error)
+	GetApplicant(string) (*Applicant, error)
```
Comment (adwells-ds) [inline go/dataspike/dataspike.go:17]: `UUID`

**Comment by adwells-ds** [comment]
Please also add short description and link to API documentation for each endpoint

File: go/dataspike/dataspike.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+	return &res, nil
+}
+
+func (dc *dataspikeClient) doRequest(method, url string, body io.Reader) ([]byte, error) {
+	req, err := http.NewRequest(method, url, body)
+	if err != nil {
+		return nil, err
+	}
+	req.Header = map[string][]string{
```
Comment (adwells-ds) [inline go/dataspike/dataspike.go:245]: Lets avoid creating even more maps on each request
```
req.Header.Set(k,v)
```

File: go/dataspike/dataspike.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+
+	return &applicant, nil
+}
+
+func (dc *dataspikeClient) CreateApplicant(applicant *ApplicantCreate) (string, error) {
+	b, err := json.Marshal(applicant)
+	if err != nil {
+		return "", err
+	}
+	fmt.Println(string(b))
```
Comment (adwells-ds) [inline go/dataspike/dataspike.go:202]:
```suggestion

```

File: go/dataspike/dataspike.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated && resp.StatusCode != http.StatusNoContent {
+		return nil, fmt.Errorf("dataspike error: %s; body: %s", resp.Status, string(bodyBytes))
+	}
+
+	return bodyBytes, nil
+}
+
+func NewClient(client *http.Client, url, token string) IDataspikeClient {
```
Comment (adwells-ds) [inline go/dataspike/dataspike.go:267]: I suggest to remove `url` from parameters available for end users, it should be internal property, users just can decide, should they use Sandbox or Production client mode

File: go/dataspike/dataspike.go [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+
+type IDataspikeClient interface {
+	GetVerificationByID(string) (*Verification, error)
+	GetVerificationByShortID(string) (*Verification, error)
+	GetApplicant(string) (*Applicant, error)
+	LinkTelegramProfile(string, string) error
+	UploadDocument(*DocumentUpload) (*Document, error)
+	CancelVerification(string) error
+	ProceedVerification(string) error
+	GetApplicantByExternal(string) (*Applicant, error)
```
Comment (adwells-ds) [inline go/dataspike/dataspike.go:22]: `GetApplicantByExternalID`

File: go/dataspike/models.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+	Dob         string `json:"dob"`
+	Gender      string `json:"gender"`
+	Citizenship string `json:"citizenship"`
+	Address     string `json:"address"`
+	Addresses   struct {
+		Residence Residence `json:"residence"`
+	} `json:"addresses"`
+}
+
+type Residence struct {
```
Comment (narma) [inline go/dataspike/models.go:91]:
```suggestion
type DetailAddress struct {
```

**Review by narma — CHANGES_REQUESTED** [review]

File: go/dataspike/dataspike.go [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+
+const tokenHeader = "ds-api-token"
+
+type IDataspikeClient interface {
+	GetVerificationByID(uuid.UUID) (*Verification, error)
+	GetVerificationByShortID(string) (*Verification, error)
+	GetApplicantByID(uuid.UUID) (*Applicant, error)
+	LinkTelegramProfile(string, string) error
+	UploadDocument(*DocumentUpload) (*Document, error)
+	CancelVerification(string) error
```
Comment (adwells-ds) [inline go/dataspike/dataspike.go:22]: UUID?)

File: go/dataspike/dataspike.go [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+	ProceedVerification(string) error
```
Comment (adwells-ds) [inline go/dataspike/dataspike.go:23]: UUID?)

File: go/dataspike/dataspike.go [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+	GetApplicantByExternalID(string) (*Applicant, error)
```
Comment (adwells-ds) [inline go/dataspike/dataspike.go:24]: UUID?)

**Review by adwells-ds — APPROVED** [review]

**Pull request merged**